### ISSUE TITLE
Shodan API key error with metasploit

### ISSUE THREAD

**Issue by surfer663** [rapid7/metasploit-framework#13598]
## Steps to reproduce

How'd you do it?

1. ...  msfconsole search shodan use auxiliary/gather/shodan_search 
2. ... set SHODAN_APIKEY set QUERY run

This section should also tell us any relevant information about the
environment; for example, if an exploit that used to work is failing,
tell us the victim operating system and service versions.

## Expected behavior it should scan

What should happen? supposed to give results of scan

## Current behavior - gives error

What happens instead?  error message : Auxiliary aborted due to failure: bad-config: 401 Unauthorized. Your SHODAN_APIKEY is invalid

I HAVE A SHODAN PAID MEMBERSHIP

You might also want to check the last ~1k lines of
`/opt/metasploit/apps/pro/engine/config/logs/framework.log` or
`~/.msf4/logs/framework.log` for relevant stack traces

### Metasploit version

Get this with the `version` command in msfconsole (or `git log -1 --pretty=oneline` for a source install).   Processing version.rc for ERB directives.
resource (version.rc)> version
Framework: 5.0.91-dev
Console  : 5.0.91-dev

### I installed Metasploit with:
- [ x] Kali package via apt  YES
- [ ] Omnibus installer (nightly)
- [ ] Commercial/Community installer (from http://www.rapid7.com/products/metasploit/download.jsp)
- [ x] Source install (please specify ruby version)    ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-linux-gnu]

### OS

What OS are you running Metasploit on?   Linux kali 5.6.0-kali2-amd64 #1 SMP Debian 5.6.14-1kali1 (2020-05-25) x86_64 GNU/Linux

**Comment by h00die** [comment]
link to module in question: https://github.com/rapid7/metasploit-framework/blob/master/modules/auxiliary/gather/shodan_search.rb

It looks like they have an officially supported ruby gem: https://github.com/picatz/shodanz linked from their website: https://developer.shodan.io/api/clients.  Maybe this module should be re-created to use the gem instead?

**Comment by surfer663** [comment]
i cant understand what changed cos right till march 2020 it worked perfectly from msfconsole.
please help me get this right as i am a newbie any help will be appreciated

**Comment by surfer663** [comment]
i forgot to mention ... when i start msfconsole i get this message

The pg and/or activerecord gem version has changed, meaning deprecated pg constants
may no longer be in use, so try deleting this file to see if the
'The PGconn, PGresult, and PGError constants are deprecated...' message has gone:
/usr/share/metasploit-framework/lib/pg/deprecated_constants.rb

**Comment by h00die** [comment]
This was resolved in a recent commit to the framework.  It most likely hasn't made it to a packaged version, and through to distros like Kali.  Either run a recent git version, or wait a week or two and the package managers will pick it up.

**Comment by adfoster-r7** [comment]
@surfer663 I've just tested this against the latest version of Metasploit with a fresh Shodan account, and I've grabbed the API key the Shodan account overview page and everything works great:

```
msf5 auxiliary(gather/shodan_search) > set SHODAN_APIKEY REMOVED_FOR_EXAMPLE
SHODAN_APIKEY => REMOVED_FOR_EXAMPLE
msf5 auxiliary(gather/shodan_search) > set query foo
query => foo
msf5 auxiliary(gather/shodan_search) > run

[*] Total: 13183461 on 131835 pages. Showing: 1 page(s)
[*] Collecting data, please wait...

Search Results
==============

 IP:Port               City                       Country         Hostname
 -------               ----                       -------         --------
 xxx.xxx.xxx.xxx:yyyy  etc etc                    etc         
 ... etc ...
```

However, if I enter in an invalid key I am able to replicate your issue:

```
msf5 auxiliary(gather/shodan_search) > set SHODAN_APIKEY wrong_key
SHODAN_APIKEY => wrong_key
msf5 auxiliary(gather/shodan_search) > set query foo
query => foo
msf5 auxiliary(gather/shodan_search) > run
\
[-] Auxiliary aborted due to failure: bad-config: 401 Unauthorized. Your SHODAN_APIKEY is invalid
[*] Auxiliary module execution completed
```

@surfer663 Could you confirm that this module isn't working with your latest valid API key from shodan?

**Comment by surfer663** [comment]
i just reset my shodan apikey 3 times and all 3 times i still got that error message

**Comment by h00die** [comment]
could you provide the entire console output from `msfconsole` through `run`.  For your key, just replace a bunch of characters.  Hopefully we'll be able to detect the issue this way.

**Comment by surfer663** [comment]
msfconsole
                                                  

 ______________________________________________________________________________
|                                                                              |
|                          3Kom SuperHack II Logon                             |
|______________________________________________________________________________|
|                                                                              |
|                                                                              |
|                                                                              |
|                 User Name:          [   security    ]                        |
|                                                                              |
|                 Password:           [               ]                        |
|                                                                              |
|                                                                              |
|                                                                              |
|                                   [ OK ]                                     |
|______________________________________________________________________________|
|                                                                              |
|                                                       https://metasploit.com |
|______________________________________________________________________________|


       =[ metasploit v5.0.92-dev                          ]
+ -- --=[ 2026 exploits - 1102 auxiliary - 343 post       ]
+ -- --=[ 562 payloads - 45 encoders - 10 nops            ]
+ -- --=[ 7 evasion                                       ]

Metasploit tip: Use sessions -1 to interact with the last opened session

msf5 > search shodan

Matching Modules
================

   #  Name                                            Disclosure Date  Rank    Check  Description
   -  ----                                            ---------------  ----    -----  -----------
   0  auxiliary/gather/shodan_honeyscore                               normal  No     Shodan Honeyscore Client
   1  auxiliary/gather/shodan_search                                   normal  No     Shodan Search
   2  auxiliary/scanner/http/influxdb_enum                             normal  No     InfluxDB Enum Utility
   3  auxiliary/scanner/http/smt_ipmi_49152_exposure  2014-06-19       normal  No     Supermicro Onboard IPMI Port 49152 Sensitive File Exposure


msf5 > use auxiliary/gather/shodan_search 
msf5 auxiliary(gather/shodan_search) > set SHODAN_APIKEY 
SHODAN_APIKEY => 
msf5 auxiliary(gather/shodan_search) > set QUERY 
QUERY => 
msf5 auxiliary(gather/shodan_search) > run

[-] Auxiliary aborted due to failure: bad-config: 401 Unauthorized. Your SHODAN_APIKEY is invalid
[*] Auxiliary module execution completed
msf5 auxiliary(gather/shodan_search) >

**Comment by adfoster-r7** [comment]
> bad-config: 401 Unauthorized. Your SHODAN_APIKEY is invalid

This response is actually from Shodan - and Shodan is giving the response that your API is invalid. It's not a problem with Metasploit itself, and I don't think there's anything we can do to help unfortunately.

@surfer663 Here's the URL that the module is visiting behind the scenes:

```
https://api.shodan.io/shodan/host/search?query=foo&page=0&key=YOUR_KEY_HERE
```

If you replace the value of `YOUR_KEY_HERE` in the above URL, what do you see? When I enter an invalid key, I get this:

```
401 Unauthorized
This server could not verify that you are authorized to access the document you requested. Either you supplied the wrong credentials (e.g., bad password), or your browser does not understand how to supply the credentials required.
```

Do you get a similar error?

**Comment by surfer663** [comment]
hi adfoster i get this similar error

[-] Auxiliary aborted due to failure: bad-config: 401 Unauthorized. Your SHODAN_APIKEY is invalid
[*] Auxiliary module execution completed

**Comment by surfer663** [comment]
and yes if i use that above url with my apikey i get same error as u did for incorrect apikey

**Comment by h00die** [comment]
I'm unable to reproduce as well.  Only thing I can offer of value is to copy/paste the API key.  I tried to transpose it, and failed on my first attempt, their 0 (zero) looks like the same height as an o (oh) with that font.

```
msf5 > use auxiliary/gather/shodan_search
msf5 auxiliary(gather/shodan_search) > set query test
query => test
msf5 auxiliary(gather/shodan_search) > set SHODAN_APIKEY aAAAaAaaaaaaaaAAAAAAaaaaAAaaaAaA
SHODAN_APIKEY => aAAAaAaaaaaaaaAAAAAAaaaaAAaaaAaA
msf5 auxiliary(gather/shodan_search) > run

[*] Total: 332479 on 3325 pages. Showing: 1 page(s)
[*] Collecting data, please wait...

Search Results
==============

 IP:Port                City                    Country                Hostname
 -------                ----                    -------                --------
```

**Comment by surfer663** [comment]
hoodie thanks for trying but i also copy and pasted made sure there is no space infront or at end but still get that message, i emailed shodan quiet a few times and they say my account is good there is no problem on their side i must contact metasploit cos problem might be there so now i dont know what else to do

**Comment by h00die** [comment]
lets try this.
load up the module, put in your info, hit run to get the 401 error.
Then type: `irb`, `datastore['SHODAN_APIKEY'].length`.

It should output similar to:
```
msf5 auxiliary(gather/shodan_search) > irb
[*] Starting IRB shell...
[*] You are in auxiliary/gather/shodan_search

>> datastore['SHODAN_APIKEY'].length
=> 32
```

@adfoster-r7 I don't see any validation on the API key performed in the module.  Looks to be `/[a-zA-Z\d]{32}/`.  Maybe one solution is adding a key validator to ensure there's some sort of reasonableness to it.

**Comment by surfer663** [comment]
according to shodan they tried with my api key and it works so now that means now problem from shodan side

**Comment by h00die** [comment]
did you try my suggestion on checking the key length?

**Comment by surfer663** [comment]
yes it says 32

**Comment by h00die** [comment]
I'm honestly out of ideas.  Other than sharing your key with us, we're not able to reproduce the issue, and everything seems correct.

**Comment by surfer663** [comment]
hoodie wont that be against shodan terms and conditions

**Comment by surfer663** [comment]
hoodie when u use ur apikey with msfconsole search does it work

**Comment by h00die** [comment]
mine worked fine, i was able to get results. not sure on terms and conditions, i'm just out of debugging ideas.

**Comment by surfer663** [comment]
does msfconsole work with ruby when searching in shodan

**Comment by surfer663** [comment]
hi hoodie & adfoster if i put my apikey in this url 
https://api.shodan.io/shodan/host/search?query=foo&page=0&key=

it works fine but not in msfconsole does that make sense

**Comment by covcov** [comment]
Hi all,

I am having the same issue, I am using shodan via metasploit like I have done some many times, haven't used it for about 6 months and now all I get is.

Auxiliary aborted due to failure: bad-config: 401 Unauthorized. Your SHODAN_APIKEY is invalid.

I have been in contact with shodan support and tried their recommendations but still have the same issue, if anyone can help I would be grateful.

If it helps here are my replies from shodan support.

Make sure you're running the latest version of the CLI:

sudo pip install -U shodan

And make sure you're not using a free API key as that might limit what you can do.


Thanks in advance 👍🙂

**Comment by NaveenGokavarapu19** [comment]
i am also facing the same issue. So i would very grateful if someone provides resolution for this error

**Comment by NaveenGokavarapu19** [comment]
Have tried to intitalize shodan from command line. The api key worked there but not working with only metasploit. So i think there is a problem with metasploit

**Comment by covcov** [comment]
Sorry should have said, I and using termux on android to run metasploit and shodan, NaveenGokavarapu19 if it's an issue with metasploit then I think I owe shodan support an apology  😱

**Comment by surfer663** [comment]
so there are others having same issues, im a bit relieved thought there is something wrong on my side only

**Comment by adfoster-r7** [comment]
@surfer663 Do you have a Metasploit slack account? I can reach out to you there to debug this further

**Comment by surfer663** [comment]
hi adfoster i dont have a slack account

**Comment by h00die** [comment]
of note, today while testing #13734 , i had a metasploit console that had been up for several days.  It had all the right info in it, as i had used it in the previous comment. I ran `run` and got a 401.  I quit MSF, restarted it, used the up key to put my API key back in, and it ran fine.  Not sure how or why that may be related, but it may help someone debug.

**Comment by rohit7771** [comment]
Hey! i am also having same problem and getting "bad-config: 401 Unauthorized. Your SHODAN_APIKEY is invalid" this error . Did anyone found what is wrong? is this the problem from metasploit or shodan?

**Comment by h00die** [comment]
@rohit7771 
What version of msf are you running?
Unfortunately the msf devs haven't been able to replicate the issue

**Comment by rohit7771** [comment]
First i tried with ,
Framework: 5.0.78-dev
Console  : 5.0.78-dev
then it showed the error so i updated msf to ,
Framework: 5.0.98-dev
Console  : 5.0.98-dev
but the error was same.

**Comment by surfer663** [comment]
hey guys im still waiting patiently for a solution lol

**Comment by adfoster-r7** [comment]
@surfer663 Would you be comfortable creating a Metasploit slack account and reaching out to me there? It might help eliminate this guess work faster.

There's a link to create new accounts here: https://metasploit.com/ within the `Join Us On` slack button.

If all goes well, I can put up a PR to fix whatever's broken, or update this issue with more details after we sync up :+1:

**Comment by surfer663** [comment]
hi adfoster i created an account on slack what do i do now

**Comment by bcoles** [comment]
> 
> 
> hi adfoster i created an account on slack what do i do now

The invitation to the Slack channel can be found here:

* https://www.metasploit.com/help

Most conversation happens in the `#general` channel.

**Comment by surfer663** [comment]
> > hi adfoster i created an account on slack what do i do now
> 
> The invitation to the Slack channel can be found here:
> 
> * https://www.metasploit.com/help
> 
> Most conversation happens in the `#general` c

**Comment by adfoster-r7** [comment]
@surfer663 I tried searching for `surfer663` on the Metasploit Slack and couldn't find it, what handle are you currently using?

**Comment by surfer663** [comment]
try darin whats ur handle on slack could not find u as well

**Comment by surfer663** [comment]
hi adfoster when wil u be on slack

**Comment by adfoster-r7** [comment]
I've sent a PM to `darin` on slack :+1:

**Comment by adfoster-r7** [comment]
The issue is now fixed.

I've also updated this module to support logging to help debug any future issues with this module faster:

```
msf5 > use auxiliary/gather/shodan_search
msf5 auxiliary(gather/shodan_search) > 
msf5 auxiliary(gather/shodan_search) > set SHODAN_APIKEY aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
SHODAN_APIKEY => aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
msf5 auxiliary(gather/shodan_search) > set query foo
query => foo
msf5 auxiliary(gather/shodan_search) > set HttpTrace true
HttpTrace => true
msf5 auxiliary(gather/shodan_search) > run

####################
# Request:
####################
GET /shodan/host/search?query=foo&key=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa&page=1 HTTP/1.1
Host: api.shodan.io
User-Agent: Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)
Content-Type: application/x-www-form-urlencoded


####################
# Response:
####################
HTTP/1.1 401 Unauthorized
Server: nginx
Date: Tue, 21 Jul 2020 17:37:48 GMT
Content-Type: text/plain; charset=UTF-8
Content-Length: 254
Connection: keep-alive
Access-Control-Allow-Origin: *

401 Unauthorized

This server could not verify that you are authorized to access the document you requested.  Either you supplied the wrong credentials (e.g., bad password), or your browser does not understand how to supply the credentials required.

[-] Auxiliary aborted due to failure: bad-config: 401 Unauthorized. Your SHODAN_APIKEY is invalid
[*] Auxiliary module execution completed
```

**Comment by surfer663** [comment]
thanks to adfoster and the other guys who helped on sorting this issue really appreciate it guys

**Comment by joethemango2018** [comment]
where can i download the updated module

**Comment by GXMX** [comment]
> O problema foi corrigido.
> 
> Também atualizei este módulo para oferecer suporte ao registro para ajudar a depurar quaisquer problemas futuros com este módulo mais rapidamente:
> 
> ```
> msf5 > use auxiliary/gather/shodan_search
> msf5 auxiliary(gather/shodan_search) > 
> msf5 auxiliary(gather/shodan_search) > set SHODAN_APIKEY aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
> SHODAN_APIKEY => aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
> msf5 auxiliary(gather/shodan_search) > set query foo
> query => foo
> msf5 auxiliary(gather/shodan_search) > set HttpTrace true
> HttpTrace => true
> msf5 auxiliary(gather/shodan_search) > run
> 
> ####################
> # Request:
> ####################
> GET /shodan/host/search?query=foo&key=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa&page=1 HTTP/1.1
> Host: api.shodan.io
> User-Agent: Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)
> Content-Type: application/x-www-form-urlencoded
> 
> 
> ####################
> # Response:
> ####################
> HTTP/1.1 401 Unauthorized
> Server: nginx
> Date: Tue, 21 Jul 2020 17:37:48 GMT
> Content-Type: text/plain; charset=UTF-8
> Content-Length: 254
> Connection: keep-alive
> Access-Control-Allow-Origin: *
> 
> 401 Unauthorized
> 
> This server could not verify that you are authorized to access the document you requested.  Either you supplied the wrong credentials (e.g., bad password), or your browser does not understand how to supply the credentials required.
> 
> [-] Auxiliary aborted due to failure: bad-config: 401 Unauthorized. Your SHODAN_APIKEY is invalid
> [*] Auxiliary module execution completed
> ```

Hey whats up? I have had the same problem for a long time and I have no idea what to do, I saw that you managed to solve the surfer, could you help me? Awaiting return.

**Comment by adfoster-r7** [comment]
@GXMX Please use the latest release of Metasploit Framework 6, if you are having issues - please raise a separate bug report :+1:

**Comment by GXMX** [comment]
Hi, good night, sorry for the delay, I only saw your message today, as soon as I wake up tomorrow I'll do that, I'm from Brazil. Thank you.

Obter o Outlook para iOS<https://aka.ms/o0ukef>
________________________________
De: adfoster-r7 <notifications@github.com>
Enviado: Tuesday, November 24, 2020 9:56:20 AM
Para: rapid7/metasploit-framework <metasploit-framework@noreply.github.com>
Cc: GXMX <vitinho11d@live.com>; Mention <mention@noreply.github.com>
Assunto: Re: [rapid7/metasploit-framework] Shodan API key error with metasploit (#13598)


@GXMX<https://github.com/GXMX> Please use the latest release of Metasploit Framework 6, if you are having issues - please raise a separate bug report 👍

**Comment by 14kaitou** [comment]
@GXMX 
Shodan API key where invalid you have to use the registered API key to use shodan it's not shodan nor the Metasploit  that's why the ouput said ``Auxiliary aborted due to failure: bad-config: 401 Unauthorized. Your SHODAN_APIKEY is invalid``

**Comment by h00die** [comment]
> @GXMX
> Shodan API key where invalid you have to use the registered API key to use shodan it's not shodan nor the Metasploit that's why the ouput said `Auxiliary aborted due to failure: bad-config: 401 Unauthorized. Your SHODAN_APIKEY is invalid`

Yes, but there was also a bug where if your shodan results contained 401 in them, it would also display this error even though your key was valid.  That's why @adfoster-r7 recommended updating, to ensure it wasn't that bug.

**Comment by 14kaitou** [comment]
e.g my API key >> ```7JBbKrmBZY*********```

**Comment by GXMX** [comment]
> @GXMX
> Shodan API key where invalid you have to use the registered API key to use shodan it's not shodan nor the Metasploit that's why the ouput said `Auxiliary aborted due to failure: bad-config: 401 Unauthorized. Your SHODAN_APIKEY is invalid`

Actually he was picking up normally, I used the key a few times and after testing what he wanted I stopped using them, when I came back this error ended up appearing, then I was a few months without using it, I'll see first if it's the update, I'm updating you here.

**Comment by GXMX** [comment]
> @GXMX Please use the latest release of Metasploit Framework 6, if you are having issues - please raise a separate bug report +1

I went to update the metasploit and after 100% I gave an 'msfconsole - version' and it appeared "Could not find reline-0.1.5 in any of the sources.
Run `bundle install` to install missing gems. ''
 I ran the bundle install and the response was "Couldn't find Gemfile". From what I saw it is something that still needs to be installed and needs to be in the correct right, but I don't know which directory to go to  :// 
 Have any idea what it is?
